specs/agents/metrics.md

```diff
@@ -50,11 +50,109 @@ Agents should not send the `system.process.cgroup.memory.mem.limit.bytes` metric
+It is periodically reported by sending the difference of the counter-value now vs its value at the time of the previous report.
+In order to save elasticsearch disk space, zero-increases are not sent.
+
+Example:
+| **Time** | **Counter** | **Reported Metric Value** |
+|----------|-------------|---------------------------|
+| 0s       | 7           | 7                         |
+| 30s      | 10          | 3                         |
+| 1m       | 10          | 0 (not sent)              |
+| 1m30s    | 12          | 2                         |
+
+
+#### Event Count Metrics
+
+Agents send their data to the Intake API of the APM-server in the form of a stream of events (e.g. transactions, spans, metricsets...).
+
+The following metrics counting these events SHOULD be exposed:
+ - `agent.events.total`: `COUNTER` of events either submitted to the internal reporting queue or dropped upon submit attempt
+ - `agent.events.dropped.queue`: `COUNTER` of events where the attempt to add it to the reporting queue failed (e.g. due to a full queue)
+ - `agent.events.dropped.error`: `COUNTER` of events dropped after they previously have been successfully added to the queue.
+
+If an agent supports these metrics, the metrics MUST be captured for all events which would reach the APM-Server in an ideal, failure-free world.
+Therefore an event MUST be taken into account for these metrics if an attempt is made to submit it to the internal reporting queue, regardless of the success.
+Events dropped before this attempt (e.g. due to [sampling](tracing-sampling.md) or [transaction_max_spans](handling-huge-traces/tracing-spans-limit.md)) are NOT counted.
+
+The value of `agent.events.dropped.error` is an upper bound for the actual number of dropped events after entering the queue.
+For example, when an request to the Intake API fails, all events within this request MUST be considered as failed.
+In reality, it is possible that e.g. half of the data was actually successfully received and forwarded by the APM server.
+
+All these metrics MUST have the label `eventType`. The possible values for `eventType` are the lowercase names of the events according to the [spec](https://github.com/elastic/apm-server/tree/main/docs/spec/v2) (e.g. `transaction`, `metricset`) or the value `other`.
+
+Agents SHOULD attempt to assign the appropriate label value based on the counted event. If this would impose significant implementation overhead, the value `other` MUST be used instead.
+
+All event count metrics are enabled/disabled via the `agent_event_metrics` configuration:
+
+|                |   |
+|----------------|---|
+| Type           | `boolean` |
+| Default        | `true` |
+| Dynamic        | `true` |
+| Central config | `true` |
+
+#### Event Queue Metrics
+
+The following metrics SHOULD be recorded by agents regarding the state of the internal event reporting queue:
+
+ - `agent.events.queue.capacity`: The number of events the reporting queue can hold at maximum
```

Comment:
The Go Agent has [something like this](https://www.elastic.co/guide/en/apm/agent/go/current/configuration.html#config-api-buffer-size), but it is measured in bytes not events. Should we just not report `agent.events.queue.*` at all?

Reply:
The main goal of the `agent.events.queue.*` metrics is to display the queue usage in percent in a dashboard. I just decided to split this in absolute capacity / size values because I thought it might be nice to view the queue capacity as well.

Due to this mismatach between how queue capacity is handled in the Go Agent vs the Java Agent I would suggest we leave out `agent.events.queue.capacity` completely and directly go for percent-based usage metrics:
 * `agent.events.queue.min_size.pct`
 * `agent.events.queue.max_size.pct`

The Java agent would then provide these metrics based on the event count, the Go Agent could do the same byte-size-based. Would this be feasible?

Reply:
What about reporting the queue utilization in percent?

Reply:
I wrote this before I saw your comment. Makes sense to me.

Reply:
@JonasKunz yes, I think that should be feasible and sounds good to me.